### PR TITLE
feat: generic cnb containerizer

### PR DESCRIPTION
**Pull request by seshapad**
Signed-off-by: seshapad <seshapad@in.ibm.com>

### CODE REVIEW

File: transformer/cnbcontainerizer.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+		var cConfig artifacts.CNBMetadataConfig
+		err = a.GetConfig(artifacts.CNBMetadataConfigType, &cConfig)
+		if err != nil {
+			logrus.Errorf("unable to load config for Transformer into %T : %s", cConfig, err)
+			continue
+		}
+		if cConfig.ImageName == "" {
+			cConfig.ImageName = common.MakeStringContainerImageNameCompliant(sConfig.ServiceName)
+		}
+		a.Configs[artifacts.CNBMetadataConfigType] = cConfig
```
Comment (ashokponkumar) [inline transformer/cnbcontainerizer.go:129]: Do we need to set it back, if imagename is not in that config?

File: types/transformer/artifacts/cnb.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+// CNBMetadataConfigType is the name of the CNB config type
+const CNBMetadataConfigType transformertypes.ConfigType = "CNBMetadata"
+
+// ProjectPath will be used as context
+
+// CNBMetadataConfig stores the configurations related to CNB
+type CNBMetadataConfig struct {
+	CNBBuilder string `json:"CNBBuilder" yaml:"CNBBuilder"`
+	ImageName  string `json:"ImageName,omitempty" yaml:"ImageName,omitempty"`
```
Comment (ashokponkumar) [inline types/transformer/artifacts/cnb.go:37]: Lets remove this from this struct to fit with the latest implementation where we put the imagename as a separate config in most cases.

File: types/transformer/artifacts/cnb.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package artifacts
+
+import (
+	transformertypes "github.com/konveyor/move2kube/types/transformer"
+)
+
+// CNBDetectedServiceArtifactType is the name of the CNB detected service artifact
+const CNBDetectedServiceArtifactType transformertypes.ArtifactType = "CNBDetectedService"
```
Comment (ashokponkumar) [inline types/transformer/artifacts/cnb.go:24]: I think we are not using this type. If so, lets remove it.

File: transformer/cnbcontainerizer.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+			irService.Containers = []core.Container{serviceContainer}
+			ir.Services[sConfig.ServiceName] = irService
+		}
+		a.Configs[irtypes.IRConfigType] = ir
+		tArtifacts = append(tArtifacts, transformertypes.Artifact{
+			Name:    a.Name,
+			Type:    artifacts.CNBMetadataArtifactType,
+			Paths:   a.Paths,
+			Configs: a.Configs,
+		})
```
Comment (ashokponkumar) [inline transformer/cnbcontainerizer.go:171]: Lets split it into the two types we had discussed before. similar to https://github.com/konveyor/move2kube/blob/d31523bde478955c3c790c4bb6c4a17156ee73b5/transformer/dockerfilegenerator/golangdockerfiletransformer.go#L173 and https://github.com/konveyor/move2kube/blob/d31523bde478955c3c790c4bb6c4a17156ee73b5/transformer/dockerfilegenerator/golangdockerfiletransformer.go#L182

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
@seshapad Awesome work bringing back the CNB containerizer. Lets plan to merge this in the next couple of days with the clean ups we had discussed.

File: transformer/cnbcontainerizer.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+		return err
+	}
+	envInfo := environment.EnvInfo{
+		Name:        tc.Name,
+		ProjectName: t.Env.GetProjectName(),
+		Source:      t.Env.GetEnvironmentSource(),
+	}
+	t.CNBEnv, err = environment.NewEnvironment(envInfo, nil, environmenttypes.Container{
+		Image:      t.CNBConfig.BuilderImageName,
+		WorkingDir: filepath.Join(string(filepath.Separator), "tmp"),
```
[the remark this answers is not in the document]
Comment (seshapad) [reply transformer/cnbcontainerizer.go]: done

File: transformer/cnbcontainerizer.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+			continue
+		}
+		var cConfig artifacts.CNBMetadataConfig
+		err = a.GetConfig(artifacts.CNBMetadataConfigType, &cConfig)
+		if err != nil {
+			logrus.Errorf("unable to load config for Transformer into %T : %s", cConfig, err)
+			continue
+		}
+		if cConfig.ImageName == "" {
+			cConfig.ImageName = common.MakeStringContainerImageNameCompliant(sConfig.ServiceName)
```
[the remark this answers is not in the document]
Comment (seshapad) [reply transformer/cnbcontainerizer.go]: done

File: transformer/cnbcontainerizer.go (hunk 1 above)
Comment (seshapad) [reply transformer/cnbcontainerizer.go]: removed it. Not required.

File: types/transformer/artifacts/cnb.go (hunk 2 above)
Comment (seshapad) [reply types/transformer/artifacts/cnb.go]: Removed this config altogether. Just using ImageName config now.

File: types/transformer/artifacts/cnb.go (hunk 3 above)
Comment (seshapad) [reply types/transformer/artifacts/cnb.go]: done

File: transformer/cnbcontainerizer.go (hunk 4 above)
Comment (seshapad) [reply transformer/cnbcontainerizer.go:165]: We just have IR in the config. Nothing else. I have retained only one artifact type as before.

File: transformer/kubernetes/clusterselector.go [hunk 7]
```diff
@@ -29,6 +29,10 @@ import (
 )
 
 const (
+	// clusterTypeKey is the key for QA ID
+	clusterTypeKey = "clustertype"
+	// defaultClusterSelectorSuffix defines the default suffix for the QA IDs
+	defaultClusterSelectorSuffix = "cs"
```
Comment (ashokponkumar) [inline transformer/kubernetes/clusterselector.go:35]: Do we need a default one? What if we make it empty by default?

**Review by ashokponkumar — CHANGES_REQUESTED** [review]
Nice work! Minor suggestions..

File: transformer/cloudfoundrytransformer.go [hunk 8]
```diff
@@ -143,6 +149,15 @@ func (t *CloudFoundry) DirectoryDetect(dir string) (services map[string][]transf
 	return services, nil
 }
 
+// resourceListToStringMap stringifies resource request
+func (t *CloudFoundry) resourceListToStringMap(rList core.ResourceList) map[string]string {
```
[the remark this answers is not in the document]
Comment (seshapad) [reply transformer/cloudfoundrytransformer.go:153]: removed

File: transformer/kubernetes/clusterselector.go (hunk 7 above)
Comment (seshapad) [reply transformer/kubernetes/clusterselector.go]: done

**Review by ashokponkumar — APPROVED** [review]
Nice!

**Pull request merged**